tracer/src/Datadog.Trace.Tools.Runner/Utils.cs

```diff
@@ -251,30 +253,39 @@ public static async Task<bool> CheckAgentConnectionAsync(string agentUrl)
 
             var tracerSettings = new TracerSettings(configurationSource);
             var settings = tracerSettings.Build();
-            var discoveryService = DiscoveryService.Create(settings.Exporter);
-
-            var agentWriter = new ApmAgentWriter(settings, new CISampler(), discoveryService);
-
-            try
-            {
-                if (!await agentWriter.Ping().ConfigureAwait(false))
+            var discoveryService = new DiscoveryService(
+                AgentTransportStrategy.Get(
+                    settings.Exporter,
+                    productName: "discovery",
+                    tcpTimeout: TimeSpan.FromSeconds(5),
+                    AgentHttpHeaderNames.MinimalHeaders,
+                    () => new MinimalAgentHeaderHelper(),
+                    uri => uri),
+                10,
+                1000,
+                int.MaxValue);
+
+            var tcs = new TaskCompletionSource<AgentConfiguration>(TaskCreationOptions.RunContinuationsAsynchronously);
+            discoveryService.SubscribeToChanges(
+                aCfg =>
                 {
-                    WriteError($"Error connecting to the Datadog Agent at {tracerSettings.Exporter.AgentUri}.");
-                    return false;
-                }
-            }
-            catch (Exception ex)
-            {
-                WriteError($"Error connecting to the Datadog Agent at {tracerSettings.Exporter.AgentUri}.");
-                AnsiConsole.WriteException(ex);
-                return false;
+                    tcs.TrySetResult(aCfg);
+                    discoveryService.DisposeAsync();
+                });
+
+            var cts = new CancellationTokenSource();
+            cts.CancelAfter(5000);
+            using (cts.Token.Register(
+                       token =>
+                       {
+                           WriteError($"Error connecting to the Datadog Agent at {tracerSettings.Exporter.AgentUri}.");
+                           tcs.TrySetResult(null);
+                           discoveryService.DisposeAsync();
+                       },
+                       cts.Token))
```

Comment:
Is the `cts.Token` state used? I ask so that we can simplify the registration code here

Reply:
ohh you're right, I was using it but I changed my mind, I forgot to remove that.

Reply:
done in [367b2bb](https://github.com/DataDog/dd-trace-dotnet/pull/3433/commits/367b2bb84fffa22d73186c47c9a62c715a90a90a)